sync: correct timestamp discontinuities in sync instead of reader

jstebbins opened the pull request:
This patch passes discontinuity information through the pipeline till it
reaches sync.c.  The timestamps are passed through the pipeline as read
and unmodified to sync.c (instead of attempting to correct
discontinuities in reader).  In sync, when we see a discontinuity,
we know where the next timestamp should be based on the timestamp
and duration of the previous buffer (before the discontinuity). So
we calculate an "SCR" offset based on the timestamp after the
discontinuity and what we calculate it should be.

The old discontinuity handling code was broken due to the following.

The MPEG STD timing model relies heavily on the decoder having an STC
that is phase lock looped to the PCRs in the stream.  When decoding a
broadcast stream, the decoder can count on the time measure between PCRs
using the STC to match to a high degree of accuracy.
I.e. STC - lastSTC  == PCR - lastPCR.  When a discontinuity occurs, the
decoder calculates a new PCR offset = PCR - STC.  I.e. the offset is the
new PCR value minus what it would have been if there had been no
discontinuity.

The above does not work without a reliable STC, which we do not have.
We have been attempting to approximate one by avereraging the duration
of received packets and extrapolating an "STC" based on the last PTS and
the average packet duration.  But this is highly variable and
unreliable.

jstebbins committed:
sync: correct timestamp discontinuities in sync instead of reader

This patch passes discontinuity information through the pipeline till it
reaches sync.c.  The timestamps are passed through the pipeline as read
and unmodified to sync.c (instead of attempting to correct
discontinuities in reader).  In sync, when we see a discontinuity,
we know where the next timestamp should be based on the timestamp
and duration of the previous buffer (before the discontinuity). So
we calculate an "SCR" offset based on the timestamp after the
discontinuity and what we calculate it should be.

The old discontinuity handling code was broken due to the following.

The MPEG STD timing model relies heavily on the decoder having an STC
that is phase lock looped to the PCRs in the stream.  When decoding a
broadcast stream, the decoder can count on the time measure between PCRs
using the STC to match to a high degree of accuracy.
I.e. STC - lastSTC  == PCR - lastPCR.  When a discontinuity occurs, the
decoder calculates a new PCR offset = PCR - STC.  I.e. the offset is the
new PCR value minus what it would have been if there had been no
discontinuity.

The above does not work without a reliable STC, which we do not have.
We have been attempting to approximate one by avereraging the duration
of received packets and extrapolating an "STC" based on the last PTS and
the average packet duration.  But this is highly variable and
unreliable.

bradleysepos commented on libhb/decavcodec.c:
```diff
@@ -519,35 +410,65 @@ static int decavcodecaWork( hb_work_object_t * w, hb_buffer_t ** buf_in,
+                                   in->data + pos, in->size - pos,
+                                   pts, pts, 0 );
+            parser_pts = pv->parser->pts;
+            pts = AV_NOPTS_VALUE;
```
Cast?

pts = (int64_t)AV_NOPTS_VALUE;

jstebbins replied:
Not needed in this case since pts is int64_t.  It's only needed when the variable is a float or double.

jstebbins committed:
decavcodec: fix data type of next_pts

It needs to be double so that partial ticks are not lost

bradleysepos commented on libhb/deccc608sub.h:
```diff
@@ -70,6 +70,7 @@ struct eia608
     int ssa_counter; // Number of subs currently written
     int screenfuls_counter; // Number of meaningful screenfuls written
     int64_t current_visible_start_ms; // At what time did the current visible buffer became so?
+    int64_t current_visible_scr_sequence; // At what time did the current visible buffer became so?
```
I'm not sure whether the comment here is intended or a copy-paste error.

jstebbins replied:
It actually was intended, though maybe I can improve the wording.   I'll change it to say SCR instead of time.

"Time" is now defined by the combination of start and scr_sequence (until a buffer reaches sync.c).  A change in the scr_sequence number indicates a discontinuity in "start" times.

jstebbins committed:
deccc608sub: clarify comment

bradleysepos commented:
At this point I can say that I've read the entire patchset and understood <= 37% of it. 😸 

Nice work, John.

jstebbins commented:
If you see something you would like more explanation of, please ask.  I'm sure there are places that could use better comments.  I usually start spotting these a couple months after initially writing new code when I'm trying to remember what it was that I did :tongue:

jstebbins committed:
sync: allow queueing more audio

Audio is small, and there is often a significant amount of audio in the
stream before the first video frame.

jstebbins committed:
sync: improve handling of damaged streams

When data is missing, the audio decoder was extrapolating timestamps
from the last pts before the error caused by the missing data which
caused sync issues.

Also, missing data can cause the video decoder to output a frame out of
order with the wrong scr sequence.  Drop such frames.

jstebbins commented:
I'm pretty happy with this patch set at this point.  I've done all the testing I can think of to do.  So unless there are more comments or someone wants to do some testing of their own, I would like to get this committed soonish.

bradleysepos commented:
Seems stable to me.
